GH-2740: Faster parsing of RDF/XML

arne-bdt opened the pull request:
Faster parsing of RDF/XML by avoiding duplicated resolving of IRIs and adding cache for IRIx in parsers
(Parsers: RRX.RDFXML_SAX, RRX.RDFXML_StAX_ev, RRX.RDFXML_StAX_sr )

GitHub issue resolved #2740

Pull request Description:
- added "public Node createURI(IRIx iriX, ...);" to the ParserProfile, which simply uses the given IRI instead of resolving it again.
- adding general IRIx caching (org.apache.jena.atlas.lib.cache.CacheSimple) in the parsers where the already cached org.apache.jena.riot.system.ParserProfileStd#resolver is not applicable
- removed unused code and variables from ParserRRX_StAX_SR and ParserRRX_StAX_EV
- added `org.apache.jena.riot.lang.rdfxml.TestXMLParser` in jena-benchmarks-jmh


Benchmark results:
```
Benchmark                                                       (param0_GraphUri)  (param1_ParserLang)  Mode  Cnt   Score   Error  Units
TestXMLParser.parseXML                                   ../testing/citations.rdf       RRX.RDFXML_SAX  avgt    5  47,232 ± 0,778   s/op
TestXMLParser.parseXML                                   ../testing/citations.rdf   RRX.RDFXML_StAX_ev  avgt    5  76,502 ± 4,390   s/op
TestXMLParser.parseXML                                   ../testing/citations.rdf   RRX.RDFXML_StAX_sr  avgt    5  48,689 ± 2,224   s/op
TestXMLParser.parseXML                                   ../testing/citations.rdf      RRX.RDFXML_ARP1  avgt    5  86,298 ± 2,440   s/op
TestXMLParser.parseXML                                ../testing/BSBM/bsbm-5m.xml       RRX.RDFXML_SAX  avgt    5   9,576 ± 0,402   s/op
TestXMLParser.parseXML                                ../testing/BSBM/bsbm-5m.xml   RRX.RDFXML_StAX_ev  avgt    5  11,562 ± 0,535   s/op
TestXMLParser.parseXML                                ../testing/BSBM/bsbm-5m.xml   RRX.RDFXML_StAX_sr  avgt    5   9,406 ± 0,465   s/op
TestXMLParser.parseXML                                ../testing/BSBM/bsbm-5m.xml      RRX.RDFXML_ARP1  avgt    5  19,738 ± 1,526   s/op
TestXMLParser.parseXML          CGMES_v2.4.15_RealGridTestConfiguration_EQ_V2.xml       RRX.RDFXML_SAX  avgt    5   0,998 ± 0,223   s/op
TestXMLParser.parseXML          CGMES_v2.4.15_RealGridTestConfiguration_EQ_V2.xml   RRX.RDFXML_StAX_ev  avgt    5   1,325 ± 0,093   s/op
TestXMLParser.parseXML          CGMES_v2.4.15_RealGridTestConfiguration_EQ_V2.xml   RRX.RDFXML_StAX_sr  avgt    5   0,985 ± 0,018   s/op
TestXMLParser.parseXML          CGMES_v2.4.15_RealGridTestConfiguration_EQ_V2.xml      RRX.RDFXML_ARP1  avgt    5   2,357 ± 0,163   s/op
TestXMLParser.parseXML         CGMES_v2.4.15_RealGridTestConfiguration_SSH_V2.xml       RRX.RDFXML_SAX  avgt    5   0,146 ± 0,029   s/op
TestXMLParser.parseXML         CGMES_v2.4.15_RealGridTestConfiguration_SSH_V2.xml   RRX.RDFXML_StAX_ev  avgt    5   0,192 ± 0,007   s/op
TestXMLParser.parseXML         CGMES_v2.4.15_RealGridTestConfiguration_SSH_V2.xml   RRX.RDFXML_StAX_sr  avgt    5   0,140 ± 0,016   s/op
TestXMLParser.parseXML         CGMES_v2.4.15_RealGridTestConfiguration_SSH_V2.xml      RRX.RDFXML_ARP1  avgt    5   0,309 ± 0,098   s/op
TestXMLParser.parseXMLJena510                            ../testing/citations.rdf       RRX.RDFXML_SAX  avgt    5  57,690 ± 0,932   s/op
TestXMLParser.parseXMLJena510                            ../testing/citations.rdf   RRX.RDFXML_StAX_ev  avgt    5  84,579 ± 4,109   s/op
TestXMLParser.parseXMLJena510                            ../testing/citations.rdf   RRX.RDFXML_StAX_sr  avgt    5  56,949 ± 0,815   s/op
TestXMLParser.parseXMLJena510                            ../testing/citations.rdf      RRX.RDFXML_ARP1  avgt    5  82,940 ± 0,815   s/op
TestXMLParser.parseXMLJena510                         ../testing/BSBM/bsbm-5m.xml       RRX.RDFXML_SAX  avgt    5  13,280 ± 0,458   s/op
TestXMLParser.parseXMLJena510                         ../testing/BSBM/bsbm-5m.xml   RRX.RDFXML_StAX_ev  avgt    5  14,994 ± 0,803   s/op
TestXMLParser.parseXMLJena510                         ../testing/BSBM/bsbm-5m.xml   RRX.RDFXML_StAX_sr  avgt    5  13,132 ± 0,166   s/op
TestXMLParser.parseXMLJena510                         ../testing/BSBM/bsbm-5m.xml      RRX.RDFXML_ARP1  avgt    5  19,125 ± 1,044   s/op
TestXMLParser.parseXMLJena510   CGMES_v2.4.15_RealGridTestConfiguration_EQ_V2.xml       RRX.RDFXML_SAX  avgt    5   1,311 ± 0,018   s/op
TestXMLParser.parseXMLJena510   CGMES_v2.4.15_RealGridTestConfiguration_EQ_V2.xml   RRX.RDFXML_StAX_ev  avgt    5   1,693 ± 0,021   s/op
TestXMLParser.parseXMLJena510   CGMES_v2.4.15_RealGridTestConfiguration_EQ_V2.xml   RRX.RDFXML_StAX_sr  avgt    5   1,332 ± 0,179   s/op
TestXMLParser.parseXMLJena510   CGMES_v2.4.15_RealGridTestConfiguration_EQ_V2.xml      RRX.RDFXML_ARP1  avgt    5   2,305 ± 0,280   s/op
TestXMLParser.parseXMLJena510  CGMES_v2.4.15_RealGridTestConfiguration_SSH_V2.xml       RRX.RDFXML_SAX  avgt    5   0,194 ± 0,028   s/op
TestXMLParser.parseXMLJena510  CGMES_v2.4.15_RealGridTestConfiguration_SSH_V2.xml   RRX.RDFXML_StAX_ev  avgt    5   0,227 ± 0,016   s/op
TestXMLParser.parseXMLJena510  CGMES_v2.4.15_RealGridTestConfiguration_SSH_V2.xml   RRX.RDFXML_StAX_sr  avgt    5   0,194 ± 0,025   s/op
TestXMLParser.parseXMLJena510  CGMES_v2.4.15_RealGridTestConfiguration_SSH_V2.xml      RRX.RDFXML_ARP1  avgt    5   0,291 ± 0,039   s/op
```

----

 - [x] Tests are included.
 -  no Documentation change and updates needed
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

afs commented on jena-arq/src/main/java/org/apache/jena/riot/RDFParserBuilder.java:
```diff
@@ -83,7 +82,6 @@ public class RDFParserBuilder {
     // HTTP
     private String appAcceptHeader = null;
     private Map<String, String> httpHeaders = new HashMap<>();
-    private HttpClient httpClient = null;
```
What's going on here?

It is important to be able to have custom HTTP setup because some SPARQL endpoints need specific settings.

arne-bdt replied:
`httpClient` was only private field in `RDFParserBuilder` with no setter or way to modify or intercept it. 
It was always only initialized by:
`HttpClient clientJDK = ( httpClient != null ) ? httpClient : HttpEnv.getDftHttpClient();`
Since `HttpEnv` only initializes a static variable `HttpEnv.httpClient` once, storing a copy in the RDFParserBuilder is not necessary to prevent duplicated initialization.

In `RDFParserBuilder#build` it was passed to 'RDFParser' where there was only one private 'field' with no setter, no way to modfy, access or intercept it.
The only usage was in `RDFParserBuilder#openTypedInputStream(String urlStr, Path path)`, so I simply moved the initialization with `HttpEnv.getDftHttpClient()` there. 
That way, the initialization, wich took about 84ms, is only be performed, when a httpClient is needed. For reading files, this is typically not the case.

afs replied:
I'd like to undertand what the cost actually is. You said "initialization" but I'm not sure what that refers to. if it parser specific or general? Which initialization step?

The Jena 5.1 error is that setter lines in `RDFParserBuilder` are commented out. 

Setting a custom `HttpClient` is important for external authentication mechanisms and complex setups. Jena does not cover all possible cases; it can't- it may involved extrat dependencies. Jena wraps up common patterns but leaving the ability to drop down to the base level.

(`HttpClient` is linked in the javadoc for `httpHeader` setting)

arne-bdt replied:
![image](https://github.com/user-attachments/assets/1a773941-e165-47b8-9778-beb5daa6b397)
In this example you see, that for reading a small xml file from disk, which takes about 420ms the initialization of the httpClient, which is not used in that case, takes 108ms of that time.

> The Jena 5.1 error is that setter lines in RDFParserBuilder are commented out.

So, the setter `RDFParserBuilder#httpClient()`, which is currently commented out in Jena 5.2.0-SNAPSHOT should be made available again? 
Idea: We could pass the probably not initialized `httpBuilder` on to the `RDFParser` and deal there with the case that it might be null by reusing the initialization code which was used in the `RDFParserBuilder` before: `( httpClient != null ) ? httpClient : HttpEnv.getDftHttpClient();`. What do you think?

> (HttpClient is linked in the javadoc for httpHeader setting)

Which javadoc exactly do you mean? I could not find it.

afs replied:
> > (HttpClient is linked in the javadoc for httpHeader setting)
> 
> Which javadoc exactly do you mean? I could not find it.

https://github.com/afs/jena/blob/452c8d287a930bb15474f3c30071aa572b27e87d/jena-arq/src/main/java/org/apache/jena/riot/RDFParserBuilder.java#L273

arne-bdt replied:
> > > (HttpClient is linked in the javadoc for httpHeader setting)
> > 
> > 
> > Which javadoc exactly do you mean? I could not find it.
> 
> https://github.com/afs/jena/blob/452c8d287a930bb15474f3c30071aa572b27e87d/jena-arq/src/main/java/org/apache/jena/riot/RDFParserBuilder.java#L273

I see now. This setter is currrently commented out in the main branch for Jena 5.2.0-SNAPSHOT and the javadoc differs. 
I will apply the above mentioned solution in this PR.

afs commented on jena-core/src/main/java/org/apache/jena/util/JenaXMLInput.java:
```diff
@@ -146,9 +147,9 @@ public static XMLInputFactory initXMLInputFactory(XMLInputFactory xmlInputFactor
         setXMLInputFactoryProperty(xmlInputFactory, XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, Boolean.FALSE);
 
-        // Not supported by Woodstox. IS_SUPPORTING_EXTERNAL_ENTITIES = false is enough.
+        // Not supported by Woodstox an Aalto. IS_SUPPORTING_EXTERNAL_ENTITIES = false is enough.
```
```suggestion
        // Not supported by Woodstox or Aalto. IS_SUPPORTING_EXTERNAL_ENTITIES = false is enough.
```

arne-bdt replied:
Thanks. I made a new commit and squashed it.

afs commented on jena-arq/src/test/java/org/apache/jena/system/TestReadXML.java:
```diff
@@ -72,10 +72,19 @@ public class TestReadXML {
+        boolean isWoodstox = name.startsWith("com.ctc.wstx.stax.");
+        boolean isAalto = name.startsWith("com.fasterxml.aalto.");
+        if(!isWoodstox && !isAalto) {
+            // Not supported by Woodstox an Aalto. IS_SUPPORTING_EXTERNAL_ENTITIES = false is enough.
```
```suggestion
            // Not supported by Woodstox or Aalto. IS_SUPPORTING_EXTERNAL_ENTITIES = false is enough.
```

arne-bdt replied:
Thanks. I made a new commit and squashed it.

arne-bdt commented:
- undid removal of httpClient from  `org.apache.jena.riot.RDFParserBuilde `
- made org.apache.jena.riot.RDFParserBuilder#httpHeader available again by uncommenting it
- moved initalization of `httpClient ` with  `HttpEnv.getDftHttpClient()` from `org.apache.jena.riot.RDFParserBuilder#build` to `org.apache.jena.riot.RDFParser#openTypedInputStream` in the case when the `RDFParserBuilder` has not been initialized with a custom httpClient. That way, it HttpEnv is not necessarily initialized when reading files but only when needed.
- updated PR and issue descriptions accordingly

arne-bdt commented:
@afs 
Should I maybe split this PR in two and make one part for the caching and one for the HTTPEnv initialization?

rvesse commented on jena-arq/src/main/java/org/apache/jena/riot/RDFParserBuilder.java:
```diff
@@ -278,13 +277,16 @@ public RDFParserBuilder httpHeader(String header, String value) {
-//        return this;
-//    }
+    /**
+     * Set an HTTP header. Any previous setting is lost.
```
```suggestion
     * Set an HTTP client. Any previous setting is lost.
```

arne-bdt replied:
Did it in the other PR.

rvesse commented on jena-arq/src/main/java/org/apache/jena/riot/lang/rdfxml/SysRRX.java:
```diff
@@ -29,9 +29,22 @@
+     * @param xmlInputFactory
+     */
+    public static <E extends XMLInputFactory> E initAndConfigure(final E xmlInputFactory) {
+         JenaXMLInput.initXMLInputFactory(xmlInputFactory);
```
```suggestion
        JenaXMLInput.initXMLInputFactory(xmlInputFactory);
```
Misaligned indentation

rvesse reviewed (commented):
Yes I think it would be worth splitting this PR

The `HttpClient` delayed init seems like a quick and independent win whereas the RDF/XML parsing and caching changes are more involved and warrants separate review IMO

afs commented:
> @afs Should I maybe split this PR in two and make one part for the caching and one for the HTTPEnv initialization?

Yes please.

afs commented:
Small request - could you put a blank line after the first line of the commit message? It separates the "subject" (the `%s` in log formatting) from the body (`%b`), otherwise the subject is the whole message.

https://git-scm.com/docs/git-commit#_discussion

afs commented:
The http client delayed setup looks OK but, as Rob says, it is orthogonal. I thought it might require an `AtomicReference` and the usual double locking pattern in `HttpEnv.getDftHttpClient` but the way here that utilizes class loading for locking looks OK (this is assuming the costs are indeed the java classloading of the networking code).

afs reviewed (approved):
Looks good and it is faster!

Just a couple of incidental javadoc suggestions.

afs commented on jena-arq/src/main/java/org/apache/jena/riot/lang/rdfxml/SysRRX.java:
```diff
@@ -29,8 +29,21 @@
 
+    /**
+     * Creates and initializes a javax.xml.stream.XMLInputFactory#newDefaultFactory().
+     * @return
```
```suggestion
     * @return XMLInputFactory
```

arne-bdt replied:
I recognized, that I kept using `javax.xml.stream.XMLInputFactory#newInstance()` instead of `javax.xml.stream.XMLInputFactory#newDefaultFactory()`, so I fixed the javadoc accordingly.

arne-bdt commented on jena-arq/src/main/java/org/apache/jena/riot/lang/rdfxml/SysRRX.java:
```diff
@@ -29,8 +29,21 @@
+    }
+
+    /**
+     * Configures the parser to be safe and sets necessary properties.
+     * This method should be called when another factory than javax.xml.stream.XMLInputFactory#newDefaultFactory()
+     * shall be used.
+     * @param xmlInputFactory
```
I recognized, that I kept using `javax.xml.stream.XMLInputFactory#newInstance()` instead of `javax.xml.stream.XMLInputFactory#newDefaultFactory()`, so I fixed the javadoc accordingly.
